### PR TITLE
api: add Self-protection framework and HTTP handler

### LINKED ISSUE #4483: Implementing the self-protection HTTP framework
## Development task
refer to #4480

### PR DESCRIPTION
**Pull request by CabinfeverB**
### What problem does this PR solve?
close #4483

This PR is going to support HTTP middleware for self-protection framework

### What is changed and how it works?

Considering HTTP and gRPC, it's better to handle the self-protection mechanism overlaying gRPC and HTTP.
As we know, PD has lots of services for other components. So The granularity of the self-protection mechanism must be service level, even service-component level. 
Based on the above considerations
> 1. need a map to store different service handlers.
> 2. HTTP and gRPC have different own actions to extract information for the service handler.

According to HTTP API, the identification is the URL path. But one URL path may have one more method for access to different services. Joining URL and method together to identify service is not harmonious.
> register mux router for the service and get router name to find the service self-protection handler

Then support middleware for HTTP
> `selfProtector` is used to implement negroni.Handler
> add `selfProtector` middleware into server handlers by negroni

### Check List

Tests

- Integration test

Code changes

- Has HTTP API interfaces change (Don't forget to [add the declarative for API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
> `http.StatusTooManyRequests` HTTP error code will be added

Side effects

- Possible performance regression

### Release note

```release-note
PD self-protection mechanism
```

### CODE REVIEW

File: tests/server/api/api_test.go [hunk 1]
```diff
@@ -110,6 +110,38 @@ func (s *serverTestSuite) TestReconnect(c *C) {
...
+	cleanup func()
+	cluster *tests.TestCluster
+}
+
+func (s *testSelfProtectorSuite) SetUpSuite(c *C) {
+	ctx, cancel := context.WithCancel(context.Background())
+	server.EnableZap = true
+	s.cleanup = cancel
+	cluster, err := tests.NewTestCluster(ctx, 3, func(conf *config.Config, serverName string) {
+		conf.TickInterval = typeutil.Duration{Duration: 50 * time.Millisecond}
```
Comment (rleungx) [inline tests/server/api/api_test.go:125]: Why change them?

File: tests/server/api/api_test.go [hunk 2]
```diff
@@ -110,6 +110,38 @@ func (s *serverTestSuite) TestReconnect(c *C) {
...
+
+func (s *testSelfProtectorSuite) TearDownSuite(c *C) {
+	s.cleanup()
+	s.cluster.Destroy()
+}
+
+func (s *testSelfProtectorSuite) TestSelfProtect(c *C) {
+	leader := s.cluster.GetServer(s.cluster.GetLeader())
+	header := mustRequestSuccess(c, leader.GetServer())
+	c.Assert(header.Get("self-protection"), Equals, "ok")
```
Comment (rleungx) [inline tests/server/api/api_test.go:142]: Can we use failpoint?

File: pkg/apiutil/serverapi/middleware.go [hunk 3]
```diff
@@ -79,6 +79,27 @@ func IsServiceAllowed(s *server.Server, group server.ServiceGroup) bool {
...
+
+func (protector *selfProtector) ServeHTTP(w http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	handler := protector.s.GetSelfProtectionHandler()
+	if handler == nil || handler.HandleHTTPSelfProtection(r) {
+		// One line below is used for middleware testing only
+		w.Header().Add("self-protection", "ok")
+
+		next(w, r)
+	} else {
+		http.Error(w, http.StatusText(http.StatusTooManyRequests), http.StatusTooManyRequests)
```
Comment (rleungx) [inline pkg/apiutil/serverapi/middleware.go:99]: StatusTooManyRequests may not cover all cases.

File: server/self_protection.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+import (
+	"net/http"
+
+	"github.com/tikv/pd/pkg/apiutil"
+)
+
+// SelfProtectionHandler is a framework to handle self protection mechanism
+// Self-protection granularity is a logical service
+type SelfProtectionHandler struct {
+	// ServiceHandlers a
```
Comment (rleungx) [inline server/self_protection.go:26]: Seem this comment is not completed.

File: server/server.go [hunk 5]
```diff
@@ -668,6 +671,11 @@ func (s *Server) stopRaftCluster() {
 	s.cluster.Stop()
 }
 
+// GetAddr returns the server urls for clients.
```
Comment (rleungx) [inline server/server.go:674]: GetSelfProtectionHandler

File: server/self_protection.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+		return true
+	}
+
+	serviceHandler, ok := h.ServiceHandlers[serviceName]
+	// if there is no service handler, go on process
+	if !ok {
+		return true
+	}
+
+	httpHandler := &HTTPServiceSelfProtectionHandler{
```
Comment (rleungx) [inline server/self_protection.go:52]: What is the purpose for creating a `HTTPServiceSelfProtectionHandler`?

File: server/server_test.go [hunk 7]
```diff
@@ -271,9 +271,5 @@ func (s *testServerHandlerSuite) TestMuxRouterName(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(resp.StatusCode, Equals, http.StatusOK)
 	c.Assert(err, IsNil)
```
Comment (rleungx) [inline server/server_test.go:273]: duplicated?

File: pkg/apiutil/apiutil.go [hunk 8]
```diff
@@ -127,3 +128,14 @@ func ErrorResp(rd *render.Render, w http.ResponseWriter, err error) {
 		rd.JSON(w, http.StatusInternalServerError, err.Error())
 	}
 }
+
+// GetHTTPRouteName return mux route name registered for ServiceName
+func GetHTTPRouteName(req *http.Request) (string, bool) {
+	route := mux.CurrentRoute(req)
+	if route != nil {
+		if route.GetName() != "" {
```
Comment (disksing) [inline pkg/apiutil/apiutil.go:136]:
```suggestion
	if route != nil && route.GetName() != "" {
```

File: server/self_protection.go [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+	"net/http"
+
+	"github.com/tikv/pd/pkg/apiutil"
+)
+
+// SelfProtectionManager is a framework to handle self protection mechanism
+// Self-protection granularity is a logical service
+type SelfProtectionManager struct {
+	// ServiceHandlers is a map to store handler owned by different services
+	ServiceHandlers map[string]*serviceSelfProtectionHandler
```
Comment (disksing) [inline server/self_protection.go:27]: no need to export

File: server/self_protection.go [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+// NewSelfProtectionManager returns a new SelfProtectionManager with config
+func NewSelfProtectionManager(server *Server) *SelfProtectionManager {
+	handler := &SelfProtectionManager{
+		ServiceHandlers: make(map[string]*serviceSelfProtectionHandler),
+	}
+	return handler
```
Comment (disksing) [inline server/self_protection.go:35]:
```suggestion
	return &SelfProtectionManager{
		ServiceHandlers: make(map[string]*serviceSelfProtectionHandler),
	}
```

File: server/self_protection.go [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+		return true
+	}
+
+	serviceHandler, ok := h.ServiceHandlers[serviceName]
+	// if there is no service handler, go on process
+	if !ok {
+		return true
+	}
+
+	httpHandler := &HTTPServiceSelfProtectionManager{
```
Comment (disksing) [inline server/self_protection.go:52]: The manager is created for every request, which looks not very good. But I think we can discuss it again in following PR.

File: server/self_protection.go [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+		req:     req,
+		handler: serviceHandler,
+	}
+	return httpHandler.Handle()
+}
+
+// ServiceSelfProtectionManager is a interface for define self-protection handler by service granularity
```
Comment (disksing) [inline server/self_protection.go:59]:
```suggestion
// ServiceSelfProtectionManager is an interface for define self-protection handler by service granularity
```

File: server/self_protection.go [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+type ServiceSelfProtectionManager interface {
+	Handle() bool
+}
+
+// HTTPServiceSelfProtectionManager implement ServiceSelfProtectionManager to handle http
```
Comment (disksing) [inline server/self_protection.go:64]:
```suggestion
// HTTPServiceSelfProtectionManager implements ServiceSelfProtectionManager to handle http requests.
```

File: server/self_protection.go [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+type HTTPServiceSelfProtectionManager struct {
+	req     *http.Request
+	handler *serviceSelfProtectionHandler
+}
+
+// Handle implement ServiceSelfProtectionManager defined function
```
Comment (disksing) [inline server/self_protection.go:70]:
```suggestion
// Handle implements ServiceSelfProtectionManager defined function
```

File: server/self_protection.go (hunk 11 above)
Comment (CabinfeverB) [reply server/self_protection.go:52]: This is a good suggestion, and it is better to implement it with a function that takes these arguments

File: server/self_protection.go (hunk 9 above)
Comment (CabinfeverB) [reply server/self_protection.go:27]: > no need to export
done

File: pkg/apiutil/apiutil.go [hunk 15]
```diff
@@ -127,3 +128,12 @@ func ErrorResp(rd *render.Render, w http.ResponseWriter, err error) {
 		rd.JSON(w, http.StatusInternalServerError, err.Error())
 	}
 }
+
+// GetRouteName return mux route name registered for ServiceName
+func GetRouteName(req *http.Request) (string, bool) {
```
Comment (rleungx) [inline pkg/apiutil/apiutil.go:133]: Seems just one return value is enough.

File: pkg/apiutil/serverapi/middleware.go [hunk 16]
```diff
@@ -79,6 +80,31 @@ func IsServiceAllowed(s *server.Server, group server.ServiceGroup) bool {
...
+	s *server.Server
+}
+
+// NewSelfProtector handle self-protection
+func NewSelfProtector(s *server.Server) negroni.Handler {
+	return &selfProtector{s: s}
+}
+
+func (protector *selfProtector) ServeHTTP(w http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	handler := protector.s.GetSelfProtectionManager()
```
Comment (rleungx) [inline pkg/apiutil/serverapi/middleware.go:93]: Can we use something like GetHandlerByReq?

**Comment by CabinfeverB** [comment]
close this PR because changing the implementation plan

**Pull request closed without merging**